### ISSUE TITLE
Negate Boolean attribute values when preceded by "!"

### ISSUE THREAD

**Issue by mhdirkse** [ibissource/iaf#1898]
**Describe the issue**
In a comment of issue https://github.com/ibissource/iaf/issues/1885 about the Frank!Doc, a note was added about Boolean attributes. It was said that when an attribute setter is of type Boolean, then the Frank developer should be able to write

    <SomePipe boolAttribute="! ${boolParameter}"...>

This should perform the not-operator to the Boolean value of boolParameter and assign it to "boolAttribute". It was said (in the issue or through the telephone) that there are Frank configs that apply this feature already.

I have tested whether this behavior has been implemented when the old ibisdoc parsing is used. I wrote a simple adapter. The adapter has `active="! ${param}"`. I ran the adapter with the Frank!Runner and manipulated the value of param through a Java system parameter `.\start.bat -Dparam=false`. My adapter was not available in Test Pipeline.

The above test suggests that this is a feature request, not a bug, but more research is needed to verify this.

**Reporter**
Martijn Dirkse

**Screenshots**

**Configuration**
```
<Configuration
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:noNamespaceSchemaLocation="./frankdoc.xsd"
  name="default"
>
    <Adapter name="default" active="! ${param}">
        <Receiver>
            <JavaListener name="dummy"></JavaListener>
        </Receiver>
        <Pipeline firstPipe="echo">
            <Exit path="EXIT" state="success" />
            <FixedResultPipe name="echo" returnString="My ${param}" substituteVars="false">
                <Forward name="success" path="EXIT" />
            </FixedResultPipe>
        </Pipeline>
    </Adapter>
</Configuration>
```

**Input**
```
Please provide an example of the input message. 
Alternatively, provide a Ladybug report.
```

**Environment**
```
Please go to the console of your Frank, click on 'Information' and copy-paste the results here. 
```

**Additional Environment**
 - DBMS: [e.g. Oracle, MSSQL, MariaDB, MySql, PostgreSQL]
 - Browser: [e.g. Chrome, Safari, Edge, Firefox]

**Additional Context**
Jeroen made the remark in issue https://github.com/ibissource/iaf/issues/1885 that was the reason the create the present issue. I think he should investigate the statement he made before this issue is implemented.

**Comment by rvdstel** [comment]
I do not know if it is relevant, but I use this feature succesfully, but I never use a space between the ! and the {
example: active="!${param}"

**Comment by mhdirkse** [comment]
Thanks, Richard van der Stel.

I updated my example to use another pipe and to omit the space after the !. My updated config contained the line:

```
<pipe className="nl.nn.adapterframework.pipes.CsvParserPipe"
                name="parse" fileContainsHeader="!${param}">
```

I started the Frank!Runner like: `.\start.bat -Dparam=false`. In the Adapter Status page I got the following error:

```
ERROR: error initializing pipeline, Exception configuring CsvParserPipe [parse]: No fieldNames specified, and fileContainsHeader=false
```

To see whether this example makes sense, I also tried to remove the "!" sign and to start with `-Dparam=true`. Then I got no error, so the "!" sign is the only thing that was rejected about my example.

Richard van der Stel, do you have an example where using the !-sign works?

**Comment by rvdstel** [comment]
https://vcs.aws.insim.biz/ibissource/dm/ibis4titan/ibis4titan_titan/-/blob/master/src/main/configuration/titan/Configuration_TS_ArchiveDocumentStreaming_4.xml

Code snippets : 

				<sender
					active="${active.batch.TS_ArchiveDocumentStreaming_4}"
					className="nl.nn.adapterframework.extensions.cmis.CmisSender"

				<sender
					active="!${active.batch.TS_ArchiveDocumentStreaming_4}"
					className="nl.nn.adapterframework.extensions.cmis.CmisSender"

SideSpecifics_AS.properties: 
active.batch.TS_ArchiveDocumentStreaming_4=false
SideSpecifics_MIG.properties: 
active.batch.TS_ArchiveDocumentStreaming_4=true					

So in the loaded configurations we see that one pipe is active on the AS node and on the MIG node we see that the other pipe is active.

**Comment by ricardovh** [comment]
As I informed you, the active-attribute is handled differently from other attributes, the configuration is first processed by https://github.com/ibissource/iaf/blob/master/core/src/main/resources/xml/xsl/active.xsl upon loading, resulting in elements with the attribute active set to `!false` and `true` to be copied and those with a different/no value to be removed from the loaded configuration.

I'm not sure how other boolean attributes are handled internally, might be that everything not having the string `true` actually just becomes `false`, regardless of the fact that you put an `!` in front of the false. I suspect that `.\start.bat -Dparam=true`, `.\start.bat -Dparam=false` and `.\start.bat -Dparam=randomvalue` would all result in the same error in your example, as the `!` will ensure it never becomes literal `true`.

**Comment by mhdirkse** [comment]
I tested my example with `-Dpara=true`, `-Dparam=false` and `-Dparam=`. In all these cases my configuration was in error, so !${param} evaluated to false. This verifies Ricardo's statement. Richard van der Stel, I guess that in your example you are also experiencing that !${active.batch.TS_ArchiveDocumentStreaming_4} always evaluates to false. I do not know how to test that in your situation, but I trust on the test I did in my simple example.

**Comment by mhdirkse** [comment]
Implementing the !-operator for attributes other than "active" is more complicated than I initially thought. We do not want to do this. I closed the pull request I was working on, #1906. We are satisfied when the !-operator only works for "active", as has been demonstrated to work. The only remaining task is to update the Frank!Doc. The Frank!Doc has to understand the special behavior of the "active" attribute.
